src/blog/first-100k-lines.vmd

```diff
@@ -0,0 +1,415 @@
+
+
+These techniques range from determinism, to testing, to type-system techniques, to general architectural best-practices.
+
+
+And since I'm a language geek, I'll throw in some nonsense about languages!
+
+
+# If you think you're using enough assertions, you're wrong
+
+I'm not exaggerating when I say assertions are the greatest thing since sliced arrays, and have cut my debugging time by half.
+
+
+<div style="text-align: center"><a href="https://i.redd.it/e1ck7vngll811.png"><img src="/images/squirrel.jpg" width="60%"/></a></div><div style="opacity: .8; text-align: center"><a href="https://i.redd.it/e1ck7vngll811.png">In your darkest hour, the assertion arrives.</a></div>
+
+
+For those unfamiliar, an assertion is basically an if-statement with a panic:
+
+```
+if thing_that_i_assume_is_four != 4 {
+  // Prints "Thing isn't 4!" and halts the program.
+  panic("Thing isn't 4!");
+}
+```
+
+and a lot of languages have an `assert` function, so it can be written like:
+
+```
+assert(thing_that_i_assume_is_four == 4, "Thing isn't 4!");
+```
+
+
+
+Rule of thumb: Every time you _think_ something is true about your data, and it's not ensured by the type system, add an assertion that checks it.
+
+Don't just occasionally use them, _drench_ your code in them. When your code so much as sneezes, they should shake off like sand after the beach. _This is the way._
+
+
+We once had a bug in the post-parser which caused some data corruption, and _thank goodness_ it was caught by an assertion in the final stage of the compiler. [# Previously known as the Hammer stage, though now it has the more boring name of FinalAstSimplifier.] If that assertion didn't catch it, it would have made it into the LLVM-generated code, and would have taken _hours_ to track down.
+
+
+Such miracles are why the Vale compiler has an entire *1,795 assertions.*
+
+
+<slice/>
+
+
+# Lean on the Type System
+
+Even if you're using a statically typed language, you can still make youre code _more_ statically typed.
+
+
+For example, a string can serve many purposes: an ID, a first name, a URL, and so on. Make a separate wrapper class for each of these, so you don't accidentally pass an ID string into a parameter expecting a URL string.
+
+Any statically-typed language has that capability, make sure to lean on it!
+
+
+Languages like Vale and Rust are particularly strong in this regard because of their [type-state programming](http://cliffle.com/blog/rust-typestate/) [# Type-state programming is a way for the compiler to ensure that a particular class doesn't fall into an invalid state.]
+and especially Vale's [Higher RAII](https://verdagon.dev/blog/higher-raii-7drl). [# Higher RAII will make sure you never forget to call a particular function.]
+
+
+# A better way to comment
```

Comment:
Keep section titles consistently cased
"Lean on the Type System" is caps, this one isn't, "Avoid Nondeterminism" is, etc